File: modules/hydromate-wake-word/android/src/main/java/expo/modules/hydromatewakeword/HydroMateWakeWordModule.kt
package expo.modules.hydromatewakeword

import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.Promise
import expo.modules.kotlin.records.Field
import expo.modules.kotlin.records.Record

internal data class WakeWordConfiguration(
  @Field val threshold: Double = WakeWordEngine.DEFAULT_THRESHOLD,
  @Field val releaseThreshold: Double = WakeWordEngine.DEFAULT_RELEASE_THRESHOLD,
  @Field val consecutiveFrames: Int = WakeWordEngine.DEFAULT_CONSECUTIVE_FRAMES,
  @Field val cooldownMillis: Int = WakeWordEngine.DEFAULT_COOLDOWN_MILLIS,
  @Field val diagnosticIntervalMillis: Int = WakeWordEngine.DEFAULT_DIAGNOSTIC_INTERVAL_MILLIS,
  @Field val diagnosticsEnabled: Boolean = false,
  @Field val calibrationOnly: Boolean = false,
  @Field val speakerVerificationEnabled: Boolean = false,
  @Field val speakerVerificationSensitivity: String = SpeakerVerifier.DEFAULT_SENSITIVITY
) : Record

class HydroMateWakeWordModule : Module() {
  private var engine: WakeWordEngine? = null
  private var speakerEnrollmentRecorder: SpeakerEnrollmentRecorder? = null
  private var speakerVerifier: SpeakerVerifier? = null
  private var debugCandidateAudioCaptureEnabled = false

  override fun definition() = ModuleDefinition {
    Name("HydroMateWakeWord")
    Events("onWakeWordDetected", "onWakeWordDiagnostic", "onWakeWordError")

    AsyncFunction("startWakeWordListening") { configuration: WakeWordConfiguration ->
      val wakeEngine = engine ?: WakeWordEngine(
        applicationContext(),
        speakerVerifier(),
        onDetected = { score, detectedAtMillis ->
          sendEvent(
            "onWakeWordDetected",
            mapOf(
              "score" to score.toDouble(),
              "detectedAtMillis" to detectedAtMillis.toDouble()
            )
          )
        },
        onDiagnostic = { score, peakScore, consecutiveFrames, state, observedAtMillis ->
          sendEvent(
            "onWakeWordDiagnostic",
            mapOf(
              "score" to score.toDouble(),
              "peakScore" to peakScore.toDouble(),
              "consecutiveFrames" to consecutiveFrames,
              "state" to state,
              "observedAtMillis" to observedAtMillis.toDouble()
            )
          )
        },
        onError = { message ->
          sendEvent("onWakeWordError", mapOf("message" to message))
        }
      ).also {
        it.setDebugCandidateAudioCaptureEnabled(debugCandidateAudioCaptureEnabled)
        engine = it
      }

      wakeEngine.setCalibrationOnly(configuration.calibrationOnly)
      wakeEngine.configureSpeakerVerification(
        configuration.speakerVerificationEnabled,
        configuration.speakerVerificationSensitivity
      )
      wakeEngine.start(
        configuration.threshold.toFloat(),
        configuration.releaseThreshold.toFloat(),
        configuration.consecutiveFrames,
        configuration.cooldownMillis.toLong(),
        configuration.diagnosticIntervalMillis.toLong(),
        configuration.diagnosticsEnabled,
        configuration.calibrationOnly
      )
    }

    AsyncFunction("stopWakeWordListening") {
      engine?.stop()
    }

    Function("isWakeWordListening") {
      engine?.isListening() == true
    }

    Function("setDebugCandidateAudioCaptureEnabled") { enabled: Boolean ->
      val allowed = WakeWordEngine.isDebugBuild(applicationContext())
      debugCandidateAudioCaptureEnabled = enabled && allowed
      engine?.setDebugCandidateAudioCaptureEnabled(debugCandidateAudioCaptureEnabled)
      debugCandidateAudioCaptureEnabled
    }

    AsyncFunction("listDebugWakeAudioFiles") {
      WakeWordEngine.listDebugWakeAudioFiles(applicationContext())
    }

    AsyncFunction("clearDebugWakeAudioFiles") {
      WakeWordEngine.clearDebugWakeAudioFiles(applicationContext())
    }

    AsyncFunction("beginSpeakerEnrollment") {
      enrollmentRecorder().beginEnrollment()
    }

    AsyncFunction("recordSpeakerEnrollmentSample") { sampleNumber: Int, promise: Promise ->
      enrollmentRecorder().recordSample(
        sampleNumber,
        onSuccess = { sample ->
          promise.resolve(
            mapOf(
              "filename" to sample.filename,
              "path" to sample.path,
              "durationMillis" to sample.durationMillis.toDouble()
            )
          )
        },
        onError = { error ->
          promise.reject(
            "ERR_SPEAKER_ENROLLMENT_RECORDING",
            error.message ?: "Speaker enrollment recording failed",
            error
          )
        }
      )
    }

    AsyncFunction("commitSpeakerEnrollment") {
      val recorder = enrollmentRecorder()
      val verifier = speakerVerifier()
      try {
        verifier.stageProfileFromPendingSamples(recorder.pendingSampleFiles())
        val samples = recorder.commitEnrollment()
        verifier.activateStagedProfile()
        samples
      } catch (error: Exception) {
        verifier.discardStagedProfile()
        throw error
      }
    }

    AsyncFunction("ensureSpeakerProfile") {
      speakerVerifier().ensureProfile()
    }

    Function("cancelSpeakerEnrollment") {
      speakerEnrollmentRecorder?.cancelEnrollment() ?: false
    }

    OnActivityEntersBackground {
      engine?.stop()
      speakerEnrollmentRecorder?.cancelEnrollment()
    }

    OnDestroy {
      debugCandidateAudioCaptureEnabled = false
      engine?.destroy()
      engine = null
      speakerEnrollmentRecorder?.destroy()
      speakerEnrollmentRecorder = null
      speakerVerifier?.destroy()
      speakerVerifier = null
    }
  }

  private fun applicationContext() =
    appContext.reactContext?.applicationContext
      ?: throw IllegalStateException("Android application context is unavailable")

  private fun enrollmentRecorder() =
    speakerEnrollmentRecorder ?: SpeakerEnrollmentRecorder(applicationContext()).also {
      speakerEnrollmentRecorder = it
    }

  private fun speakerVerifier() =
    speakerVerifier ?: SpeakerVerifier(applicationContext()).also {
      speakerVerifier = it
    }
}

File: modules/hydromate-wake-word/src/HydroMateWakeWordModule.ts
import {
  NativeModule,
  requireOptionalNativeModule,
  type EventSubscription,
} from "expo-modules-core";
import { Platform } from "react-native";

export type WakeWordConfiguration = {
  threshold: number;
  releaseThreshold: number;
  consecutiveFrames: number;
  cooldownMillis: number;
  diagnosticIntervalMillis: number;
  diagnosticsEnabled: boolean;
  calibrationOnly: boolean;
  speakerVerificationEnabled: boolean;
  speakerVerificationSensitivity: "lenient" | "balanced" | "strict";
};

export type WakeWordDetectedEvent = {
  score: number;
  detectedAtMillis: number;
};

export type WakeWordDiagnosticEvent = {
  score: number;
  peakScore: number;
  consecutiveFrames: number;
  state: string;
  observedAtMillis: number;
};

export type SpeakerEnrollmentSample = {
  filename: string;
  path: string;
  durationMillis: number;
};

type WakeWordEvents = {
  onWakeWordDetected(event: WakeWordDetectedEvent): void;
  onWakeWordDiagnostic(event: WakeWordDiagnosticEvent): void;
  onWakeWordError(event: { message: string }): void;
};

declare class HydroMateWakeWordNativeModule extends NativeModule<WakeWordEvents> {
  startWakeWordListening(
    configuration: WakeWordConfiguration
  ): Promise<boolean>;
  stopWakeWordListening(): Promise<void>;
  isWakeWordListening(): boolean;
  setDebugCandidateAudioCaptureEnabled(enabled: boolean): boolean;
  listDebugWakeAudioFiles(): Promise<string[]>;
  clearDebugWakeAudioFiles(): Promise<number>;
  beginSpeakerEnrollment(): Promise<boolean>;
  recordSpeakerEnrollmentSample(
    sampleNumber: number
  ): Promise<SpeakerEnrollmentSample>;
  commitSpeakerEnrollment(): Promise<string[]>;
  ensureSpeakerProfile(): Promise<boolean>;
  cancelSpeakerEnrollment(): boolean;
}

const nativeModule =
  requireOptionalNativeModule<HydroMateWakeWordNativeModule>(
    "HydroMateWakeWord"
  );

export const isNativeWakeWordAvailable =
  Platform.OS === "android" && nativeModule !== null;

export async function startWakeWordListening(
  configuration: WakeWordConfiguration
) {
  if (!nativeModule) {
    return false;
  }

  return nativeModule.startWakeWordListening(configuration);
}

export async function stopWakeWordListening() {
  await nativeModule?.stopWakeWordListening();
}

export function isWakeWordListening() {
  return nativeModule?.isWakeWordListening() ?? false;
}

export function setDebugCandidateAudioCaptureEnabled(enabled: boolean) {
  return nativeModule?.setDebugCandidateAudioCaptureEnabled(enabled) ?? false;
}

export async function listDebugWakeAudioFiles() {
  return (await nativeModule?.listDebugWakeAudioFiles()) ?? [];
}

export async function clearDebugWakeAudioFiles() {
  return (await nativeModule?.clearDebugWakeAudioFiles()) ?? 0;
}

export async function beginSpeakerEnrollment() {
  if (!nativeModule) {
    throw new Error("Speaker enrollment recording is unavailable");
  }
  return nativeModule.beginSpeakerEnrollment();
}

export async function recordSpeakerEnrollmentSample(sampleNumber: number) {
  if (!nativeModule) {
    throw new Error("Speaker enrollment recording is unavailable");
  }
  return nativeModule.recordSpeakerEnrollmentSample(sampleNumber);
}

export async function commitSpeakerEnrollment() {
  if (!nativeModule) {
    throw new Error("Speaker enrollment recording is unavailable");
  }
  return nativeModule.commitSpeakerEnrollment();
}

export async function ensureSpeakerProfile() {
  if (!nativeModule) return false;
  return nativeModule.ensureSpeakerProfile();
}

export function cancelSpeakerEnrollment() {
  return nativeModule?.cancelSpeakerEnrollment() ?? false;
}

export function addWakeWordDetectedListener(
  listener: (event: WakeWordDetectedEvent) => void
): EventSubscription {
  if (!nativeModule) {
    return { remove() {} };
  }

  return nativeModule.addListener("onWakeWordDetected", listener);
}

export function addWakeWordDiagnosticListener(
  listener: (event: WakeWordDiagnosticEvent) => void
): EventSubscription {
  if (!nativeModule) {
    return { remove() {} };
  }

  return nativeModule.addListener("onWakeWordDiagnostic", listener);
}

export function addWakeWordErrorListener(
  listener: (event: { message: string }) => void
): EventSubscription {
  if (!nativeModule) {
    return { remove() {} };
  }

  return nativeModule.addListener("onWakeWordError", listener);
}
